File: Kernel Simulator/Shell/Shells/Text/Commands/QueryWord.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports KS.Misc.Reflection
Imports Extensification.IntegerExts
Imports KS.Misc.Editors.TextEdit

Namespace Shell.Shells.Text.Commands
    ''' <summary>
    ''' Queries a word in a specified line or all lines
    ''' </summary>
    ''' <remarks>
    ''' You can use this command to query a word and get its number from the specified line or all lines.
    ''' </remarks>
    Class TextEdit_QueryWordCommand
        Inherits CommandExecutor
        Implements ICommand

        Public Overrides Sub Execute(StringArgs As String, ListArgsOnly As String(), ListSwitchesOnly As String()) Implements ICommand.Execute
            If ListArgsOnly.Length = 2 Then
                If IsStringNumeric(ListArgsOnly(1)) Then
                    If CInt(ListArgsOnly(1)) <= TextEdit_FileLines.Count Then
                        Dim QueriedChars As Dictionary(Of Integer, String) = TextEdit_QueryWord(ListArgsOnly(0), ListArgsOnly(1))
                        For Each WordIndex As Integer In QueriedChars.Keys
                            Write("- {0}: ", False, ColTypes.ListEntry, WordIndex)
                            Write("{0} ({1})", True, ColTypes.ListValue, ListArgsOnly(0), TextEdit_FileLines(ListArgsOnly(1)))
                        Next
                    Else
                        Write(DoTranslation("The specified line number may not be larger than the last file line number."), True, ColTypes.Error)
                    End If
                ElseIf ListArgsOnly(1).ToLower = "all" Then
                    Dim QueriedWords As Dictionary(Of Integer, Dictionary(Of Integer, String)) = TextEdit_QueryWord(ListArgsOnly(0))
                    For Each LineIndex As Integer In QueriedWords.Keys
                        For Each WordIndex As Integer In QueriedWords(LineIndex).Keys
                            Write("- {0}:{1}: ", False, ColTypes.ListEntry, LineIndex, WordIndex)
                            Write("{0} ({1})", True, ColTypes.ListValue, ListArgsOnly(0), TextEdit_FileLines(LineIndex))
                        Next
                    Next
                End If
            ElseIf ListArgsOnly.Length > 2 Then
                If IsStringNumeric(ListArgsOnly(1)) And IsStringNumeric(ListArgsOnly(2)) Then
                    If CInt(ListArgsOnly(1)) <= TextEdit_FileLines.Count And CInt(ListArgsOnly(2)) <= TextEdit_FileLines.Count Then
                        Dim LineNumberStart As Integer = ListArgsOnly(1)
                        Dim LineNumberEnd As Integer = ListArgsOnly(2)
                        LineNumberStart.SwapIfSourceLarger(LineNumberEnd)
                        For LineNumber = LineNumberStart To LineNumberEnd
                            Dim QueriedChars As Dictionary(Of Integer, String) = TextEdit_QueryWord(ListArgsOnly(0), LineNumber)
                            For Each WordIndex As Integer In QueriedChars.Keys
                                Write("- {0}:{1}: ", False, ColTypes.ListEntry, LineNumber, WordIndex)
                                Write("{0} ({1})", True, ColTypes.ListValue, ListArgsOnly(0), TextEdit_FileLines(ListArgsOnly(1)))
                            Next
                        Next
                    Else
                        Write(DoTranslation("The specified line number may not be larger than the last file line number."), True, ColTypes.Error)
                    End If
                End If
            End If
        End Sub

    End Class
End Namespace

File: public/Nitrocid/Shell/Shells/Text/Commands/QueryWord.cs
//
// Nitrocid KS  Copyright (C) 2018-2025  Aptivi
//
// This file is part of Nitrocid KS
//
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using Nitrocid.ConsoleBase.Colors;
using Nitrocid.ConsoleBase.Writers;
using Terminaux.Writer.ConsoleWriters;
using Nitrocid.Files.Editors.TextEdit;
using Nitrocid.Kernel.Exceptions;
using Nitrocid.Languages;
using Nitrocid.Misc.Reflection;
using Nitrocid.Shell.ShellBase.Commands;
using System;
using Textify.General;

namespace Nitrocid.Shell.Shells.Text.Commands
{
    /// <summary>
    /// Queries a word in a specified line or all lines
    /// </summary>
    /// <remarks>
    /// You can use this command to query a word and get its number from the specified line or all lines.
    /// </remarks>
    class QueryWordCommand : BaseCommand, ICommand
    {

        public override int Execute(CommandParameters parameters, ref string variableValue)
        {
            if (parameters.ArgumentsList.Length == 2)
            {
                if (TextTools.IsStringNumeric(parameters.ArgumentsList[1]))
                {
                    if (Convert.ToInt32(parameters.ArgumentsList[1]) <= TextEditShellCommon.FileLines.Count)
                    {
                        int LineIndex = Convert.ToInt32(parameters.ArgumentsList[1]);
                        var QueriedChars = TextEditTools.QueryWord(parameters.ArgumentsList[0], LineIndex);
                        TextWriters.Write("- {0}: ", false, KernelColorType.ListEntry, LineIndex);

                        // Process the output
                        string text = TextEditShellCommon.FileLines[LineIndex - 1];
                        var Words = text.Split(' ');
                        for (int wordIndex = 0; wordIndex < Words.Length; wordIndex++)
                        {
                            string word = Words[wordIndex];
                            TextWriters.Write($"{(QueriedChars.Contains(wordIndex) ? KernelColorTools.GetColor(KernelColorType.Success).VTSequenceForeground : "")}{word} ", false, KernelColorType.ListValue);
                        }
                        TextWriterRaw.Write();
                        return 0;
                    }
                    else
                    {
                        TextWriters.Write(Translate.DoTranslation("The specified line number may not be larger than the last file line number."), true, KernelColorType.Error);
                        return KernelExceptionTools.GetErrorCode(KernelExceptionType.TextEditor);
                    }
                }
                else if (parameters.ArgumentsList[1].Equals("all", StringComparison.OrdinalIgnoreCase))
                {
                    var QueriedWords = TextEditTools.QueryWord(parameters.ArgumentsList[0]);
                    foreach (var QueriedWord in QueriedWords)
                    {
                        int LineIndex = QueriedWord.Item1;
                        var QueriedChars = TextEditTools.QueryWord(parameters.ArgumentsList[0], LineIndex + 1);
                        TextWriters.Write("- {0}: ", false, KernelColorType.ListEntry, LineIndex + 1);

                        // Process the output
                        string text = TextEditShellCommon.FileLines[LineIndex];
                        var Words = text.Split(' ');
                        for (int wordIndex = 0; wordIndex < Words.Length; wordIndex++)
                        {
                            string word = Words[wordIndex];
                            TextWriters.Write($"{(QueriedChars.Contains(wordIndex) ? KernelColorTools.GetColor(KernelColorType.Success).VTSequenceForeground : "")}{word} ", false, KernelColorType.ListValue);
                        }
                        TextWriterRaw.Write();
                    }
                    return 0;
                }
            }
            else if (parameters.ArgumentsList.Length > 2)
            {
                if (TextTools.IsStringNumeric(parameters.ArgumentsList[1]) & TextTools.IsStringNumeric(parameters.ArgumentsList[2]))
                {
                    if (Convert.ToInt32(parameters.ArgumentsList[1]) <= TextEditShellCommon.FileLines.Count & Convert.ToInt32(parameters.ArgumentsList[2]) <= TextEditShellCommon.FileLines.Count)
                    {
                        int LineNumberStart = Convert.ToInt32(parameters.ArgumentsList[1]);
                        int LineNumberEnd = Convert.ToInt32(parameters.ArgumentsList[2]);
                        LineNumberStart.SwapIfSourceLarger(ref LineNumberEnd);
                        for (int LineNumber = LineNumberStart; LineNumber <= LineNumberEnd; LineNumber++)
                        {
                            var QueriedChars = TextEditTools.QueryWord(parameters.ArgumentsList[0], LineNumber);
                            int LineIndex = LineNumber - 1;
                            TextWriters.Write("- {0}: ", false, KernelColorType.ListEntry, LineIndex);

                            // Process the output
                            string text = TextEditShellCommon.FileLines[LineIndex];
                            var Words = text.Split(' ');
                            for (int wordIndex = 0; wordIndex < Words.Length; wordIndex++)
                            {
                                string word = Words[wordIndex];
                                TextWriters.Write($"{(QueriedChars.Contains(wordIndex) ? KernelColorTools.GetColor(KernelColorType.Success).VTSequenceForeground : "")}{word} ", false, KernelColorType.ListValue);
                            }
                            TextWriterRaw.Write();
                        }
                        return 0;
                    }
                    else
                    {
                        TextWriters.Write(Translate.DoTranslation("The specified line number may not be larger than the last file line number."), true, KernelColorType.Error);
                        return KernelExceptionTools.GetErrorCode(KernelExceptionType.TextEditor);
                    }
                }
            }
            return 0;
        }

    }
}
